File: src/1_100/0040_combination-sum-ii/combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        recursive(0, candidates, 0, new Stack<Integer>(), target);
        return new ArrayList<List<Integer>>(res);
    }

    Set<List<Integer>> res = new HashSet<List<Integer>>();
    void recursive(int i,int[] arr,int tmp,Stack<Integer> tmp_arr,int target) {
        if(tmp==target) {
            res.add(new ArrayList(tmp_arr));
        }
        for(int j=i;j<arr.length;j++) {
            if(tmp+arr[j] <= target) {
                tmp_arr.add(arr[j]);
                recursive(j + 1, arr, tmp + arr[j], tmp_arr, target);
                //tmp -= arr[j];
                tmp_arr.pop();
            }
        }
    }
}

File: src/1_100/0040_combination-sum-ii/combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        if not candidates:
            return []
        res = []
        n = len(candidates)
        candidates = sorted(candidates)
        def dfs(index,arr,num):
            if num>target or index>n:
                return
            if num==target:
                res.append(arr)
                return
            if index<n and (candidates[index]>target or num+candidates[index]>target):
                return
            for i in xrange(index,n):
                if i>index and candidates[i]==candidates[i-1]:
                    continue
                dfs(i+1,arr+[candidates[i]],num+candidates[i])
        dfs(0,[],0)
        return res